File: src/main/java/com/browserstack/gradle/EspressoTask.java
 package com.browserstack.gradle;

import org.gradle.api.tasks.TaskAction;
import java.io.FileReader;
import java.net.HttpURLConnection;
import java.util.HashMap;
import java.util.Map;
import java.nio.file.Path;
import com.browserstack.json.JSONObject;
import com.browserstack.httputils.HttpUtils;
import org.json.simple.parser.JSONParser;
import org.gradle.api.tasks.Input;

public class EspressoTask extends BrowserStackTask {

  @Input
  private String configFilePath;

  private String testSuite;

  public void setConfigFilePath(String filePath) {
    this.configFilePath = filePath;
  }

  public String getConfigFilePath() {
    return configFilePath;
  }


  private String constructBuildParams() {
    JSONObject params = constructDefaultBuildParams();

    JSONParser jsonParser = new JSONParser();
    org.json.simple.JSONObject caps;

    try {
      Object obj = jsonParser.parse(new FileReader(getConfigFilePath()));
      org.json.simple.JSONObject jsonObject = (org.json.simple.JSONObject) obj;

      params.put("testSuite", testSuite);

      for (Object o : jsonObject.keySet()) {
        String key = (String) o;
        params.put(key, jsonObject.get(key));
      }
    } catch (Exception e) {
        System.out.println("Config file parsing failed with below error: ");
        e.printStackTrace();
    }

    return params.toString();
  }

  private void uploadTestSuite(Path testApkPath) throws Exception {
    try {
      final boolean wrapPropsAsInternal = false;
      final Map<String, String> extraProperties = new HashMap<>();
      extraProperties.put(KEY_EXTRA_CUSTOM_ID, this.customId);
      HttpURLConnection con = HttpUtils.sendPostApp(
              isDebug,
              wrapPropsAsInternal,
              getHost() + Constants.TEST_SUITE_UPLOAD_PATH,
              basicAuth(),
              testApkPath.toString(),
              extraProperties
      );
      int responseCode = con.getResponseCode();
      System.out.println("TestSuite upload Response Code : " + responseCode);

      JSONObject response = new JSONObject(HttpUtils.getResponse(con, responseCode));

      if (responseCode == 200) {
        testSuite = (String) response.get("test_suite_url");
      } else {
        throw new Exception("TestSuite upload failed");
      }

    } catch (Exception e) {
      e.printStackTrace();
      throw e;
    }
  }

  private void executeTest() throws Exception {
    try {
      HttpURLConnection con = HttpUtils.sendPostBody(
              getHost() + Constants.BUILD_PATH,
              basicAuth(),
              constructBuildParams()
      );
      int responseCode = con.getResponseCode();
      System.out.println("Response Code : " + responseCode);

      JSONObject response = new JSONObject(HttpUtils.getResponse(con, responseCode));

      if (responseCode == 200) {
          String build_id = response.getString("build_id");
          displayDashboardURL(build_id);
          return;
      }

      return;

    } catch (Exception e) {
      e.printStackTrace();
      throw e;
    }
  }

  private void displayDashboardURL(String build_id) {
    System.out.println("View build status at " + Constants.APP_AUTOMATE_HOST + "/builds/" + build_id);
  }

  public void verifyParams() throws Exception {
    String username = this.getUsername();
    String accessKey = this.getAccessKey();
    if (username == null || accessKey == null || configFilePath == null) {
      throw new Exception("`username`, `accessKey` and `configFilePath` are compulsory");
    }
  }

  @TaskAction
  void uploadAndExecuteTest() throws Exception {
    verifyParams();
    final boolean ignoreTestPath = false;
    final boolean wrapPropsAsInternal = false;
    Map<String, Path> apkFiles = locateApks(ignoreTestPath);
    uploadApp(
            wrapPropsAsInternal,
            Constants.APP_AUTOMATE_ESPRESSO_UPLOAD_PATH,
            apkFiles.get(BrowserStackTask.KEY_FILE_DEBUG)
    );
    uploadTestSuite(apkFiles.get(BrowserStackTask.KEY_FILE_TEST));
    executeTest();
  }
}

File: src/main/java/com/browserstack/gradle/EspressoTask.kt
package com.browserstack.gradle

import com.browserstack.httputils.HttpUtils
import com.browserstack.json.JSONObject
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.TaskAction
import java.nio.file.Path
import javax.inject.Inject

open class EspressoTask @Inject constructor(
    host: String,
    variantBaseName: String,
    @Input val extension: BrowserStackConfigExtension
) : BrowserStackTask(
    host = host,
    variantBaseName = variantBaseName,
    username = extension.username,
    accessKey = extension.accessKey
) {

    override fun verifyParams() {
        super.verifyParams()
        require(extension.devices.isNotEmpty()) { "You must specify at least one device (devices)" }
    }

    @TaskAction
    open fun uploadAndExecuteTest() {
        verifyParams()

        // First upload app
        val apkFiles = getApkCollection(findTestApk = true)
        val app = uploadApp(Constants.APP_AUTOMATE_UPLOAD_PATH, apkFiles.debugApk)

        // Then upload and execute test suite
        val testSuite = uploadTestSuite(apkFiles.testApk ?: error("testApk not found"))
        executeTestSuite(app, testSuite).also(::displayDashboardURL)
    }

    private fun constructBuildParams(app: String, testSuite: String): String {
        return constructDefaultBuildParams().apply {
            put("app", app)
            put("testSuite", testSuite)

            put("devices", extension.devices)
            put("class", extension.classes)
            put("package", extension.packages)
            put("size", extension.sizes)
            put("annotation", extension.annotations)
            put("otherApps", extension.otherApps)
            put("video", extension.video)
            put("deviceLogs", extension.deviceLogs)
            put("networkLogs", extension.networkLogs)
            put("local", extension.local)
            put("localIdentifier", extension.localIdentifier)
            put("networkProfile", extension.networkProfile)
            put("callbackURL", extension.callbackURL)
            put("timezone", extension.timeZone)
            put("appStoreConfiguration", extension.appStoreConfigurationMap)
            put("enableSpoonFramework", extension.enableSpoonFramework)
            put("disableAnimations", extension.disableAnimations)
            put("allowDeviceMockServer", extension.allowDeviceMockServer)
            put("customBuildName", extension.customBuildName)
            put("customBuildNotifyURL", extension.customBuildNotifyURL)
            put("project", extension.projectName)
            put("geoLocation", extension.geoLocation)
            put("language", extension.language)
            put("locale", extension.locale)
            put("deviceOrientation", extension.deviceOrientation)
        }.toString()
    }

    private fun uploadTestSuite(testApkPath: Path?): String {
        return try {
            val connection = HttpUtils.sendPost(
                host + Constants.TEST_SUITE_UPLOAD_PATH,
                basicAuth(),
                null,
                testApkPath.toString()
            )
            when (val responseCode =
                connection.responseCode.also { println("TestSuite upload Response Code: $it") }) {
                200 -> {
                    val response = JSONObject(HttpUtils.getResponse(connection, responseCode))
                    response.getString("test_url")
                }
                else -> throw Exception("TestSuite upload failed")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
    }

    private fun executeTestSuite(app: String, testSuite: String): String {
        return try {
            val connection = HttpUtils.sendPost(
                host + Constants.BUILD_PATH,
                basicAuth(),
                constructBuildParams(app, testSuite),
                null
            )
            val responseCode = connection.responseCode.also { println("Response Code : $it") }
            val response = JSONObject(HttpUtils.getResponse(connection, responseCode))
            response.getString("build_id")
        } catch (e: Exception) {
            e.printStackTrace()
            throw e
        }
    }

    private fun displayDashboardURL(buildId: String) {
        println("View build status at ${Constants.APP_AUTOMATE_HOST}/builds/$buildId")
    }
}
